README.md
# QR Code Encoder For .Net 二维码生成器.Net版

[![License](https://img.shields.io/github/license/ALI1416/QRCodeEncoder.Net?label=License)](https://www.apache.org/licenses/LICENSE-2.0.txt)
[![.Net Support](https://img.shields.io/badge/.NET%20Standard-2.0+-green)](https://openjdk.org/)
[![NuGet Gallery](https://img.shields.io/nuget/v/Z.QRCodeEncoder.Net?label=NuGet%20Gallery)](https://www.nuget.org/packages/Z.QRCodeEncoder.Net)
[![Tag](https://img.shields.io/github/v/tag/ALI1416/QRCodeEncoder.Net?label=Tag)](https://github.com/ALI1416/QRCodeEncoder.Net/tags)
[![Repo Size](https://img.shields.io/github/repo-size/ALI1416/QRCodeEncoder.Net?label=Repo%20Size&color=success)](https://github.com/ALI1416/QRCodeEncoder.Net/archive/refs/heads/master.zip)

[![DotNet CI](https://github.com/ALI1416/QRCodeEncoder.Net/actions/workflows/ci.yml/badge.svg)](https://github.com/ALI1416/QRCodeEncoder.Net/actions/workflows/ci.yml)

## 简介

本项目参考了[micjahn/ZXing.Net](https://github.com/micjahn/ZXing.Net)，只编写了生成器部分，并对处理逻辑进行了大量优化，构建后`Z.QRCodeEncoder.Net.dll`文件仅`40kb`

注意：本项目不提供二维码绘制方法，如需绘制请看`使用示例`

### 其他语言项目

- `Java` : [ALI1416/qrcode-encoder](https://github.com/ALI1416/qrcode-encoder)
- `JavaScript` : [ALI1416/qrcode-encoder-js](https://github.com/ALI1416/qrcode-encoder-js)

## 依赖导入

```sh
dotnet add package Z.QRCodeEncoder.Net --version 1.1.0
```

## 方法和参数

### 二维码 QRCode

| 参数名        | 中文名   | 类型   | 默认值     |
| ------------- | -------- | ------ | ---------- |
| content       | 内容     | string | (无)       |
| level         | 纠错等级 | int    | 0          |
| mode          | 编码模式 | int    | (自动探测) |
| versionNumber | 版本号   | int    | (最小版本) |

### 版本 Version

| 参数名        | 中文名     | 类型 | 默认值     |
| ------------- | ---------- | ---- | ---------- |
| length        | 内容字节数 | int  | (无)       |
| level         | 纠错等级   | int  | (无)       |
| mode          | 编码模式   | int  | (无)       |
| versionNumber | 版本号     | int  | (最小版本) |

### 掩模模板 MaskPattern

| 参数名  | 中文名   | 类型    |
| ------- | -------- | ------- |
| data    | 数据     | bool[]  |
| version | 版本     | Version |
| level   | 纠错等级 | int     |

### 纠错等级 level

| 值  | 等级 | 纠错率 |
| --- | ---- | ------ |
| 0   | L    | 7%     |
| 1   | M    | 15%    |
| 2   | Q    | 25%    |
| 3   | H    | 30%    |

### 编码模式 mode

| 值  | 模式             | 备注                                     |
| --- | ---------------- | ---------------------------------------- |
| 0   | NUMERIC          | 数字0-9                                  |
| 1   | ALPHANUMERIC     | 数字0-9、大写字母A-Z、符号(空格)$%*+-./: |
| 2   | BYTE(ISO-8859-1) | 兼容ASCII                                |
| 3   | BYTE(UTF-8)      |                                          |

### 版本号 versionNumber

取值范围：`[1,40]`

## 使用示例

### 生成并绘制二维码

Program.cs

```csharp
string content = "1234😀";
int level = 0;
QRCode qrCode = new QRCode(content, level);
Bitmap bitmap = ImageUtils.QrBytes2Bitmap(qrCode.Matrix, 10);
ImageUtils.SaveBitmap(bitmap, "qr.png");
```

ImageUtils.cs

```csharp
private static readonly Brush BLACK_BRUSH = new SolidBrush(Color.Black);

public static Bitmap QrBytes2Bitmap(bool[,] bytes, int pixelSize)
{
    int length = bytes.GetLength(0);
    List<Rectangle> rects = new List<Rectangle>();
    for (int x = 0; x < length; x++)
    {
        for (int y = 0; y < length; y++)
        {
            if (bytes[x, y])
            {
                rects.Add(new Rectangle((x + 1) * pixelSize, (y + 1) * pixelSize, pixelSize, pixelSize));
            }
        }
    }
    int size = (length + 2) * pixelSize;
    Bitmap bitmap = new Bitmap(size, size);
    using (Graphics g = Graphics.FromImage(bitmap))
    {
        g.FillRectangles(BLACK_BRUSH, rects.ToArray());
    }
    return bitmap;
}

public static void SaveBitmap(Bitmap bitmap, string path)
{
    bitmap.Save(path, ImageFormat.Png);
}
```

更多请见[测试](./Z.QRCodeEncoder.Net.Test)

## 更新日志

[点击查看](./CHANGELOG.md)

## 参考

- [micjahn/ZXing.Net](https://github.com/micjahn/ZXing.Net)

## 关于

<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://www.404z.cn/images/about.dark.svg">
  <img alt="About" src="https://www.404z.cn/images/about.light.svg">
</picture>


CHANGELOG.md
# 更新日志

## v1.1.0 `2023.6.9`

- 新增异常类

## v1.0.2 `2023.5.31`

- 代码优化

## v1.0.1 `2023.5.19`

- 修复可能会导致的空指针异常

## v1.0.0 `2023.5.16`

- 正式版发布


Z.QRCodeEncoder.Net/QRCode.cs
using System;
using System.Text;

namespace Z.QRCodeEncoder.Net
{

    /// <summary>
    /// 二维码
    /// <para>@createDate 2023/05/16 11:11:11</para>
    /// <para>@author ALI[ali-k@foxmail.com]</para>
    /// <para>@since 1.0.0</para>
    /// </summary>
    public class QRCode
    {

        /// <summary>
        /// 纠错等级
        /// </summary>
        public readonly int Level;
        /// <summary>
        /// 编码模式
        /// </summary>
        public readonly int Mode;
        /// <summary>
        /// 版本
        /// </summary>
        public readonly Version Version;
        /// <summary>
        /// 版本号
        /// </summary>
        public readonly int VersionNumber;
        /// <summary>
        /// 掩模模板
        /// </summary>
        public readonly MaskPattern MaskPattern;
        /// <summary>
        /// 掩模模板号
        /// </summary>
        public readonly int MaskPatternNumber;
        /// <summary>
        /// 矩阵
        /// <para>false白 true黑</para>
        /// </summary>
        public readonly bool[,] Matrix;

        /// <summary>
        /// 构造二维码
        /// <para>纠错等级 0 L 7%</para>
        /// <para>编码模式 自动探测</para>
        /// <para>版本号 最小版本</para>
        /// </summary>
        /// <param name="content">
        /// 内容
        /// </param>
        public QRCode(string content) : this(content, null, null, null) { }

        /// <summary>
        /// 构造二维码
        /// <para>编码模式 自动探测</para>
        /// <para>版本号 最小版本</para>
        /// </summary>
        /// <param name="content">
        /// 内容
        /// </param>
        /// <param name="level">
        /// 纠错等级
        /// <para>0 L 7%(默认)</para>
        /// <para>1 M 15%</para>
        /// <para>2 Q 25%</para>
        /// <para>3 H 30%</para>
        /// </param>
        public QRCode(string content, int? level) : this(content, level, null, null) { }

        /// <summary>
        /// 构造二维码
        /// <para>版本号 最小版本</para>
        /// </summary>
        /// <param name="content">
        /// 内容
        /// </param>
        /// <param name="level">
        /// 纠错等级
        /// <para>0 L 7%(默认)</para>
        /// <para>1 M 15%</para>
        /// <para>2 Q 25%</para>
        /// <para>3 H 30%</para>
        /// </param>
        /// <param name="mode">
        /// 编码模式(默认自动探测)
        /// <para>0 NUMERIC 数字0-9</para>
        /// <para>1 ALPHANUMERIC 数字0-9、大写字母A-Z、符号(空格)$%*+-./:</para>
        /// <para>2 BYTE(ISO-8859-1)</para>
        /// <para>3 BYTE(UTF-8)</para>
        /// </param>
        public QRCode(string content, int? level, int? mode) : this(content, level, mode, null) { }

        /// <summary>
        /// 构造二维码
        /// </summary>
        /// <param name="content">
        /// 内容
        /// </param>
        /// <param name="level">
        /// 纠错等级
        /// <para>0 L 7%(默认)</para>
        /// <para>1 M 15%</para>
        /// <para>2 Q 25%</para>
        /// <para>3 H 30%</para>
        /// </param>
        /// <param name="mode">
        /// 编码模式(默认自动探测)
        /// <para>0 NUMERIC 数字0-9</para>
        /// <para>1 ALPHANUMERIC 数字0-9、大写字母A-Z、符号(空格)$%*+-./:</para>
        /// <para>2 BYTE(ISO-8859-1)</para>
        /// <para>3 BYTE(UTF-8)</para>
        /// </param>
        /// <param name="versionNumber">
        /// 版本号(默认最小版本)
        /// <para>[1,40]</para>
        /// </param>
        public QRCode(string content, int? level, int? mode, int? versionNumber)
        {
            int levelValue;
            int modeValue;
            /* 数据 */
            // 纠错等级
            if (level == null)
            {
                levelValue = 0;
            }
            else if (level < 0 || level > 3)
            {
                throw new QRCodeException("纠错等级 " + level + " 不合法！应为 [0,3]");
            }
            else
            {
                levelValue = (int)level;
            }
            Level = levelValue;
            // 编码模式
            if (mode == null)
            {
                modeValue = DetectionMode(content);
            }
            else if (mode < 0 || mode > 3)
            {
                throw new QRCodeException("编码模式 " + mode + " 不合法！应为 [0,3]");
            }
            else
            {
                int detectionMode = DetectionMode(content);
                if (mode < detectionMode)
                {
                    throw new QRCodeException("编码模式 " + mode + " 太小！最小为 " + detectionMode);
                }
                modeValue = (int)mode;
            }
            Mode = modeValue;
            // 内容bytes
            byte[] contentBytes = Encoding.UTF8.GetBytes(content);
            // 版本
            Version = new Version(contentBytes.Length, Level, Mode, versionNumber);
            VersionNumber = Version.VersionNumber;
            // 数据bits
            bool[] dataBits = new bool[Version.DataBits];
            // 填充数据
            switch (Mode)
            {
                // 填充编码模式为NUMERIC的数据
                case 0:
                    {
                        ModeNumbers(dataBits, contentBytes, Version);
                        break;
                    }
                // 填充编码模式为ALPHANUMERIC的数据
                case 1:
                    {
                        ModeAlphaNumeric(dataBits, contentBytes, Version);
                        break;
                    }
                // 填充编码模式为BYTE编码格式为ISO-8859-1的数据
                case 2:
                    {
                        ModeByteIso88591(dataBits, contentBytes, Version);
                        break;
                    }
                // 填充编码模式为BYTE编码格式为UTF-8的数据
                default:
                    {
                        ModeByteUtf8(dataBits, contentBytes, Version);
                        break;
                    }
            }

            /* 纠错 */
            int[,] ec = Version.Ec;
            // 数据块数 或 纠错块数
            int blocks = 0;
            for (int i = 0; i < Version.Ec.GetLength(0); i++)
            {
                blocks += Version.Ec[i, 0];
            }
            // 纠错块字节数
            int ecBlockBytes = (Version.DataAndEcBits - Version.DataBits) / 8 / blocks;
            int[][] dataBlocks = new int[blocks][];
            int[][] ecBlocks = new int[blocks][];
            int blockNum = 0;
            int dataByteNum = 0;
            for (int i = 0; i < ec.GetLength(0); i++)
            {
                int count = ec[i, 0];
                int dataBytes = ec[i, 1];
                for (int j = 0; j < count; j++)
                {
                    // 数据块
                    int[] dataBlock = QRCodeUtils.GetBytes(dataBits, dataByteNum * 8, dataBytes);
                    dataBlocks[blockNum] = dataBlock;
                    // 纠错块
                    ecBlocks[blockNum] = ReedSolomon.Encoder(dataBlock, ecBlockBytes);
                    blockNum++;
                    dataByteNum += dataBytes;
                }
            }

            /* 交叉数据和纠错 */
            bool[] dataAndEcBits = new bool[Version.DataAndEcBits];
            int dataBlockMaxBytes = dataBlocks[blocks - 1].Length;
            int dataAndEcBitPtr = 0;
            for (int i = 0; i < dataBlockMaxBytes; i++)
            {
                for (int j = 0; j < blocks; j++)
                {
                    if (dataBlocks[j].Length > i)
                    {
                        QRCodeUtils.AddBits(dataAndEcBits, dataAndEcBitPtr, dataBlocks[j][i], 8);
                        dataAndEcBitPtr += 8;
                    }
                }
            }
            for (int i = 0; i < ecBlockBytes; i++)
            {
                for (int j = 0; j < blocks; j++)
                {
                    QRCodeUtils.AddBits(dataAndEcBits, dataAndEcBitPtr, ecBlocks[j][i], 8);
                    dataAndEcBitPtr += 8;
                }
            }

            /* 构造掩模模板 */
            MaskPattern = new MaskPattern(dataAndEcBits, Version, Level);
            MaskPatternNumber = MaskPattern.Best;
            Matrix = QRCodeUtils.Convert(MaskPattern.Patterns[MaskPatternNumber], Version.Dimension);
        }

        /// <summary>
        /// 填充编码模式为NUMERIC的数据
        /// </summary>
        /// <param name="dataBits">数据bits</param>
        /// <param name="contentBytes">内容bytes</param>
        /// <param name="version">版本</param>
        private static void ModeNumbers(bool[] dataBits, byte[] contentBytes, Version version)
        {
            // 数据指针
            int ptr = 0;
            // 模式指示符(4bit) NUMERIC 0b0001=1
            // 数据来源 ISO/IEC 18004-2015 -> 7.4.1 -> Table 2 -> QR Code symbols列Numbers行
            QRCodeUtils.AddBits(dataBits, ptr, 1, 4);
            ptr += 4;
            // 内容字节数
            int contentLength = contentBytes.Length;
            // `内容字节数`bit数(10/12/14bit)
            int contentBytesBits = version.ContentBytesBits;
            QRCodeUtils.AddBits(dataBits, ptr, contentLength, contentBytesBits);
            ptr += contentBytesBits;
            // 内容 3个字符10bit 2个字符7bit 1个字符4bit
            for (int i = 0; i < contentLength - 2; i += 3)
            {
                QRCodeUtils.AddBits(dataBits, ptr, (contentBytes[i] - 48) * 100 + (contentBytes[i + 1] - 48) * 10 + contentBytes[i + 2] - 48, 10);
                ptr += 10;
            }
            switch (contentLength % 3)
            {
                case 2:
                    {
                        QRCodeUtils.AddBits(dataBits, ptr, (contentBytes[contentLength - 2] - 48) * 10 + contentBytes[contentLength - 1] - 48, 7);
                        ptr += 7;
                        break;
                    }
                case 1:
                    {
                        QRCodeUtils.AddBits(dataBits, ptr, contentBytes[contentLength - 1] - 48, 4);
                        ptr += 4;
                        break;
                    }
            }
            // 结束符和补齐符
            TerminatorAndPadding(dataBits, version.DataBits, ptr);
        }

        /// <summary>
        /// 填充编码模式为ALPHANUMERIC的数据
        /// </summary>
        /// <param name="dataBits">数据bits</param>
        /// <param name="contentBytes">内容bytes</param>
        /// <param name="version">版本</param>
        private static void ModeAlphaNumeric(bool[] dataBits, byte[] contentBytes, Version version)
        {
            // 数据指针
            int ptr = 0;
            // 模式指示符(4bit) ALPHANUMERIC 0b0010=2
            // 数据来源 ISO/IEC 18004-2015 -> 7.4.1 -> Table 2 -> QR Code symbols列Alphanumeric行
            QRCodeUtils.AddBits(dataBits, ptr, 2, 4);
            ptr += 4;
            // 内容字节数
            int contentLength = contentBytes.Length;
            // `内容字节数`bit数(9/11/13bit)
            int contentBytesBits = version.ContentBytesBits;
            QRCodeUtils.AddBits(dataBits, ptr, contentLength, contentBytesBits);
            ptr += contentBytesBits;
            // 内容 2个字符11bit 1个字符6bit
            for (int i = 0; i < contentLength - 1; i += 2)
            {
                QRCodeUtils.AddBits(dataBits, ptr, ALPHA_NUMERIC_TABLE[contentBytes[i]] * 45 + ALPHA_NUMERIC_TABLE[contentBytes[i + 1]], 11);
                ptr += 11;
            }
            if (contentLength % 2 == 1)
            {
                QRCodeUtils.AddBits(dataBits, ptr, ALPHA_NUMERIC_TABLE[contentBytes[contentLength - 1]], 6);
                ptr += 6;
            }
            // 结束符和补齐符
            TerminatorAndPadding(dataBits, version.DataBits, ptr);
        }

        /// <summary>
        /// 填充编码模式为BYTE编码格式为ISO-8859-1的数据
        /// </summary>
        /// <param name="dataBits">数据bits</param>
        /// <param name="contentBytes">内容bytes</param>
        /// <param name="version">版本</param>
        private static void ModeByteIso88591(bool[] dataBits, byte[] contentBytes, Version version)
        {
            // 数据指针
            int ptr = 0;
            // 模式指示符(4bit) BYTE 0b0100=4
            // 数据来源 ISO/IEC 18004-2015 -> 7.4.1 -> Table 2 -> QR Code symbols列Byte行
            QRCodeUtils.AddBits(dataBits, ptr, 4, 4);
            ptr += 4;
            // 内容字节数
            int contentLength = contentBytes.Length;
            // `内容字节数`bit数(8/16bit)
            int contentBytesBits = version.ContentBytesBits;
            QRCodeUtils.AddBits(dataBits, ptr, contentLength, contentBytesBits);
            ptr += contentBytesBits;
            // 内容
            for (int i = 0; i < contentLength; i++)
            {
                QRCodeUtils.AddBits(dataBits, ptr, contentBytes[i], 8);
                ptr += 8;
            }
            // 结束符和补齐符
            TerminatorAndPadding(dataBits, version.DataBits, ptr);
        }

        /// <summary>
        /// 填充编码模式为BYTE编码格式为UTF-8的数据
        /// </summary>
        /// <param name="dataBits">数据bits</param>
        /// <param name="contentBytes">内容bytes</param>
        /// <param name="version">版本</param>
        private static void ModeByteUtf8(bool[] dataBits, byte[] contentBytes, Version version)
        {
            // 数据指针
            int ptr = 0;
            // ECI模式指示符(4bit) 0b0111=7
            // 数据来源 ISO/IEC 18004-2015 -> 7.4.1 -> Table 2 -> QR Code symbols列ECI行
            QRCodeUtils.AddBits(dataBits, ptr, 7, 4);
            ptr += 4;
            // ECI指定符 UTF-8(8bit) 0b00011010=26
            // 数据来源 ?
            QRCodeUtils.AddBits(dataBits, ptr, 26, 8);
            ptr += 8;
            // 模式指示符(4bit) BYTE 0b0100=4
            // 数据来源 ISO/IEC 18004-2015 -> 7.4.1 -> Table 2 -> QR Code symbols列Byte行
            QRCodeUtils.AddBits(dataBits, ptr, 4, 4);
            ptr += 4;
            // 内容字节数
            int contentLength = contentBytes.Length;
            // `内容字节数`bit数(8/16bit)
            int contentBytesBits = version.ContentBytesBits;
            QRCodeUtils.AddBits(dataBits, ptr, contentLength, contentBytesBits);
            ptr += contentBytesBits;
            // 内容
            for (int i = 0; i < contentLength; i++)
            {
                QRCodeUtils.AddBits(dataBits, ptr, contentBytes[i], 8);
                ptr += 8;
            }
            // 结束符和补齐符
            TerminatorAndPadding(dataBits, version.DataBits, ptr);
        }

        /// <summary>
        /// 结束符和补齐符
        /// </summary>
        /// <param name="data">数据bits</param>
        /// <param name="dataBits">数据bits数</param>
        /// <param name="ptr">数据指针</param>
        private static void TerminatorAndPadding(bool[] data, int dataBits, int ptr)
        {
            // 如果有刚好填满，则不需要结束符和补齐符
            // 如果还剩1-8bit，需要1-8bit结束符，不用补齐符
            // 如果还剩8+bit，先填充4bit结束符，再填充结束符使8bit对齐，再交替补齐符至填满
            if (dataBits - ptr > 7)
            {
                // 结束符(4bit)
                // 数据来源 ISO/IEC 18004-2015 -> 7.4.9
                ptr += 4;
                // 结束符(8bit对齐)
                ptr = (((ptr - 1) / 8) + 1) * 8;
                // 补齐符 交替0b11101100=0xEC和0b00010001=0x11至填满
                // 数据来源 ISO/IEC 18004-2015 -> 7.4.10
                int count = (dataBits - ptr) / 8;
                for (int i = 0; i < count; i++)
                {
                    if (i % 2 == 0)
                    {
                        Array.Copy(NUMBER_0xEC_8BITS, 0, data, ptr, 8);
                    }
                    else
                    {
                        Array.Copy(NUMBER_0x11_8BITS, 0, data, ptr, 8);
                    }
                    ptr += 8;
                }
            }
        }

        /// <summary>
        /// 探测编码模式
        /// </summary>
        /// <param name="content">内容</param>
        /// <returns>
        /// 编码模式
        /// <para>0 NUMERIC 数字0-9</para>
        /// <para>1 ALPHANUMERIC 数字0-9、大写字母A-Z、符号(空格)$%*+-./:</para>
        /// <para>2 BYTE(ISO-8859-1)</para>
        /// <para>3 BYTE(UTF-8)</para>
        /// </returns>
        private static int DetectionMode(string content)
        {
            int length = content.Length;
            // 为了与ZXing结果保持一致，长度为0时使用BYTE(ISO-8859-1)编码
            if (length == 0)
            {
                return 2;
            }
            // BYTE(UTF-8)
            for (int i = 0; i < length; i++)
            {
                if (content[i] > 255)
                {
                    return 3;
                }
            }
            // BYTE(ISO-8859-1)
            for (int i = 0; i < length; i++)
            {
                if (content[i] > 127 || ALPHA_NUMERIC_TABLE[content[i]] > 44)
                {
                    return 2;
                }
            }
            // ALPHANUMERIC 数字0-9、大写字母A-Z、符号(空格)$%*+-./:
            for (int i = 0; i < length; i++)
            {
                if (ALPHA_NUMERIC_TABLE[content[i]] > 9)
                {
                    return 1;
                }
            }
            // NUMERIC 数字0-9
            return 0;
        }

        /// <summary>
        /// 数字0xEC
        /// <para>数据来源 ISO/IEC 18004-2015 -> 7.4.10 -> 11101100</para>
        /// </summary>
        private static readonly bool[] NUMBER_0xEC_8BITS = QRCodeUtils.GetBits(0xEC, 8);
        /// <summary>
        /// 数字0x11
        /// <para>数据来源 ISO/IEC 18004-2015 -> 7.4.10 -> 00010001</para>
        /// </summary>
        private static readonly bool[] NUMBER_0x11_8BITS = QRCodeUtils.GetBits(0x11, 8);

        /// <summary>
        /// ALPHANUMERIC模式映射表
        /// <para>数字0-9 [0x30,0x39] [0,9]</para>
        /// <para>大写字母A-Z [0x41,0x5A] [10,35]</para>
        /// <para>(空格) [0x20] [36]</para>
        /// <para>$ [0x24] [37]</para>
        /// <para>% [0x25] [38]</para>
        /// <para>* [0x2A] [39]</para>
        /// <para>+ [0x2B] [40]</para>
        /// <para>- [0x2D] [41]</para>
        /// <para>. [0x2E] [42]</para>
        /// <para>/ [0x2F] [43]</para>
        /// <para>: [0x3A] [44]</para>
        /// <para>数据来源 ISO/IEC 18004-2015 -> 7.4.5 -> Table 6</para>
        /// </summary>
        private static readonly byte[] ALPHA_NUMERIC_TABLE = new byte[]
        {
            127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, // 0x00-0x0F
            127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, // 0x10-0x1F
             36, 127, 127, 127,  37,  38, 127, 127, 127, 127,  39,  40, 127,  41,  42,  43, // 0x20-0x2F
              0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  44, 127, 127, 127, 127, 127, // 0x30-0x3F
            127,  10,  11,  12,  13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24, // 0x40-0x4F
             25,  26,  27,  28,  29,  30,  31,  32,  33,  34,  35, 127, 127, 127, 127, 127, // 0x50-0x5F
            127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, // 0x60-0x6F
            127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, 127, // 0x70-0x7F
        };

    }
}
